File: scripts/providers/nominatim.py
"""
Nominatim (OpenStreetMap) geocoding provider.

Free, no API key required. Usage policy:
  - max 1 request/second
  - identifiable User-Agent
  - https://operations.osmfoundation.org/policies/nominatim/
"""

from __future__ import annotations

import time

import requests

from scripts.providers.base import GeocodingProvider, GeocodingResult

NOMINATIM_URL = "https://nominatim.openstreetmap.org/search"
DEFAULT_RATE_LIMIT_S = 1.1  # slightly above 1s for safety
DEFAULT_TIMEOUT_S = 15


class NominatimProvider(GeocodingProvider):
    """GeocodingProvider backed by Nominatim/OpenStreetMap."""

    name = "nominatim"

    def __init__(
        self,
        session: requests.Session | None = None,
        user_agent: str | None = None,
        rate_limit_s: float = DEFAULT_RATE_LIMIT_S,
        timeout_s: int = DEFAULT_TIMEOUT_S,
    ):
        self._session = session or requests.Session()
        self._rate_limit_s = rate_limit_s
        self._timeout_s = timeout_s
        self._last_request_ts = 0.0

        if user_agent:
            self._session.headers.setdefault("User-Agent", user_agent)

    def _wait_rate_limit(self) -> None:
        elapsed = time.monotonic() - self._last_request_ts
        if elapsed < self._rate_limit_s:
            time.sleep(self._rate_limit_s - elapsed)
        self._last_request_ts = time.monotonic()

    def geocode(self, query: str) -> GeocodingResult | None:
        if not query or not query.strip():
            return None

        self._wait_rate_limit()

        try:
            response = self._session.get(
                NOMINATIM_URL,
                params={
                    "q": query,
                    "format": "jsonv2",
                    "limit": 1,
                    "countrycodes": "br",
                    "addressdetails": 0,
                },
                timeout=self._timeout_s,
            )
            if not response.ok:
                return None
            data = response.json()
            if not data:
                return None
            item = data[0]
            return GeocodingResult(
                lat=float(item["lat"]),
                lng=float(item["lon"]),
                source=self.name,
            )
        except (requests.RequestException, ValueError, KeyError):
            return None

File: lib/providers/nominatim.js
/**
 * Nominatim (OpenStreetMap) geocoding provider.
 *
 * Used as a fallback when BrasilAPI returns a CEP without coordinates
 * (which, in practice, is the vast majority of CEPs).
 *
 * Public Nominatim policy: max 1 req/s and a descriptive User-Agent.
 * We don't sleep here — at API request rate this is well under the cap,
 * and we cache results in Supabase so a repeated CEP costs nothing.
 *
 * Docs: https://nominatim.org/release-docs/develop/api/Search/
 */

const DEFAULT_TIMEOUT_MS = 8000;
const NOMINATIM_URL = 'https://nominatim.openstreetmap.org/search';

class NominatimProvider {
  constructor({ userAgent, timeoutMs = DEFAULT_TIMEOUT_MS } = {}) {
    this._userAgent =
      userAgent ||
      'hospitais-referencia-api/1.0 (+https://github.com/Codar-Sistemas/hospitais-referencia-api)';
    this._timeoutMs = timeoutMs;
  }

  get name() {
    return 'nominatim';
  }

  /**
   * Geocode a free-text Brazilian address. Returns { lat, lng } or null.
   * Caller is responsible for caching.
   */
  async geocode(query) {
    if (!query || !query.trim()) return null;

    const controller = new AbortController();
    const timer = setTimeout(() => controller.abort(), this._timeoutMs);

    try {
      const url = new URL(NOMINATIM_URL);
      url.searchParams.set('q', query);
      url.searchParams.set('format', 'json');
      url.searchParams.set('limit', '1');
      url.searchParams.set('countrycodes', 'br');
      url.searchParams.set('addressdetails', '0');

      const r = await fetch(url, {
        headers: {
          'User-Agent': this._userAgent,
          'Accept-Language': 'pt-BR,pt;q=0.9',
        },
        signal: controller.signal,
      });
      if (!r.ok) return null;
      const data = await r.json();
      if (!Array.isArray(data) || data.length === 0) return null;
      const lat = parseFloat(data[0].lat);
      const lng = parseFloat(data[0].lon);
      if (isNaN(lat) || isNaN(lng)) return null;
      return { lat, lng };
    } catch {
      return null;
    } finally {
      clearTimeout(timer);
    }
  }
}

module.exports = { NominatimProvider };
